packages/@aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts

```diff
@@ -333,7 +346,7 @@ type GraphAnnotation =
   | { readonly type: 'step'; readonly step: Step; isBuildStep?: boolean }
   | { readonly type: 'self-update' }
   | { readonly type: 'prepare'; readonly stack: StackDeployment }
-  | { readonly type: 'execute'; readonly stack: StackDeployment; readonly captureOutputs: boolean }
+  | { readonly type: 'execute'; readonly stack: StackDeployment; readonly captureOutputs: boolean; readonly withoutChangeset?: boolean }
```

Comment:
These types constitute a contract, that is consumed for example by https://github.com/cdklabs/cdk-pipelines-github

Changing what the `execute` node means will break other implementors. Instead, add a new node type (for example, `deploy`) that has a new meaning.

Reply:
Would it break, of there is new optional field? I first checked separate one, but at least thst point it seemed to generate more changes.

Of course now with working tests it would be easier to refactor.

Reply:
I think that adding new node here is easier breaking change, as node is created at https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts#L138-L142.

Here if we determine the new `deploy` node if no prepare step is needed, existing functionality would fail as e.g. GitHub engine would get different node what it can't handle at https://github.com/cdklabs/cdk-pipelines-github/blob/3fa9bb2df734c4eb6b5ce05946e728af65881037/src/pipeline.ts#L350-L351.

Or did I misunderstood? Optional prop should be invisible to all existing implementors, and allows those to continue as is.

Of course that execute node in general is a bit confusing, as with that existing logic to leave out prepare step single step is not really execute but deploy https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/pipelines/lib/helpers-internal/pipeline-graph.ts#L137.